OPEN_PROBLEMS/ENHANCED_BITSWAP_GRAPHSYNC.md

```diff
@@ -0,0 +1,128 @@
+# Enhanced Bitswap/GraphSync with more Network Smarts
+
+## Short Description
+> In one sentence or paragraph.
+
+Once IPFS has resolved the CID to a peerID and further to the exact location of the node that  stores the requested content, it then uses a very simple protocol called bitswap to exchange content between the requestor and the server.
+
+Although bitswap is simple and generally works, *its performance is significantly suboptimal*. This is mainly due to the fact that a node cannot request a subgraph of the DAG and results in many round-trips in order to “walk down” the DAG. The current operation of bitswap is also very often linked to duplicate transmission and receipt of content which overloads both the end nodes but also the network.
+
+## Long Description
+
+In order to synchronise a DAG between untrusted nodes, bitswap is exploiting the content-addressability feature of IPFS. This means that bitswap starts with the root hash of the DAG and once it has the data, it rehashes the data to verify that it results to the same hash. The node can now trust this block and hence, can continue with the rest of the blocks in the DAG (aka walk the DAG), until it gets the whole DAG and the data associated with it.
+
+In the current implementation of bitswap, requesting nodes send their WANT lists to all the peers they are directly connected to. This inevitably results in potentially duplicate traffic travelling back to the receiving node. When the receiving node has received the block(s) it asked for, it sends a CANCEL message to its peers to let them know that the block(s) is not in its WANT list anymore.
```

Comment:
```suggestion
In the current implementation of bitswap, requesting nodes send their WANT lists to all the peers they are directly connected to. This results in potentially duplicate data being sent back to the receiving node. When the receiving node has received the block(s) it asked for, it sends a CANCEL message to its peers to let them know that the block(s) is not in its WANT list anymore.
```

Reply:
To be clear, is that "(s)" correct? Do multiple cancellations get collated into the same CANCEL message? The go-bitswap readme (https://github.com/ipfs/go-bitswap#requesting-blocks) says that is indeed the case, both for requests and cancellations. However, the asynchronous nature of block retrieval (and, consequently, cancellation) would make this unlikely unless outgoing messages are queued for a _long_ (whatever that means) time. There is an implied efficiency trade-off here, between reducing outgoing signalling traffic and reducing redundant inbound traffic.